File: cpp/3461_has_same_digits/has_same_digits.cpp
#include <cassert>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  bool hasSameDigits(string s) {
    int n = s.length();
    for (int i = 1; i <= n - 2; i++) {
      for (int j = 0; j <= n - 1 - i; j++) {
        s[j] = ((s[j] - '0') + (s[j + 1] - '0')) % 10 + '0';
      }
    }
    return s[0] == s[1];
  }
};

int main() {
  vector<tuple<string, bool>> tests{
      {"3902", true},
      {"34789", false},
  };

  for (auto &[s, ans] : tests) {
    assert(Solution().hasSameDigits(s) == ans);
  }
}

File: golang/2461_has_same_digits/has_same_digits.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func hasSameDigits(s string) bool {
	n := len(s)
	arr := []byte(s)
	for i := 1; i <= n-2; i++ {
		for j := 0; j <= n-1-i; j++ {
			arr[j] = byte(((int(arr[j]-'0') + int(arr[j+1]-'0')) % 10) + '0')
		}
	}
	return arr[0] == arr[1]
}

func main() {
	tests := []struct {
		s   string
		ans bool
	}{
		{"3902", true},
		{"34789", false},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, hasSameDigits(test.s), index)
	}
}
